### ISSUE TITLE
Model::loadCounts does not work correctly when relationship contains a join

### ISSUE THREAD

**Issue by troy-whitespark** [laravel/framework#41042]
(recreating this as an issue, as requested in my pull with a failing test)

- Laravel Version: 8+
- PHP Version: 8+
- Database Driver & Version: 

### Description:

This problem stemmed from a real issue I was having building relationships that are scoped to particular "locations" in my application.

The issue is, loadCounts and withCounts both create a new subquery and copy the wheres, but they do not copy the join constraints - in the case where we're doing something like inner joining to a pivot table to get permutations of content and location pairs, that means that the relationship_count and relationship()->count() will return two different values and cause bugs.

### Steps To Reproduce:

Existing pull request with failing test is here: https://github.com/laravel/framework/pull/40867

**Comment by derekmd** [comment]
TLDR below; I'm unsure if this is worth investigating since `has()`/`doesntHave()` + `with()` eager loading do play nice with the relationship aggregate methods.

---

From your [given example test](https://github.com/laravel/framework/pull/40867/files), instead of table joins I'd suggest using `has()` / `whereHas()` when defining the relationship, then `with()`/`load()` to access nested `Location` models:

### App\Models\User

```php
class User extends Model
{
    public function locations()
    {
        return $this->hasMany(Location::class);
    }

    public function posts()
    {
        return $this->hasMany(Post::class);
    }

    public function postsOfManyLocations()
    {
        return $this->posts()
            ->where('global', '=', false)
            ->has('locations'); // WHERE EXISTS subquery is run
    }
}
```

### App\Models\Post

```php
class Post extends Model
{
    public function locations()
    {
        return $this->belongsToMany(Location::class, 'post_locations');
    }

    public function user()
    {
        return $this->belongsTo(User::class);
    }
}
```

### App\Models\Location

```php
class Location extends Model
{
    public function posts()
    {
        return $this->belongsToMany(Post::class);
    }

    public function user()
    {
        return $this->belongsTo(User::class)
    }
}
```

```php
// $user = User::withCount('postsOfManyLocations')->first();
$user = User::first()->loadCount('postsOfManyLocations');

// this will be accurate
$user->posts_of_many_locations_count;

$user->postsOfManyLocations->load('locations');
```

---

The docs may need to discourage Eloquent models using `join()` (only covered in the [general query builder section](https://laravel.com/docs/queries#joins)) since there are a few other active record issues I'm aware of:

1. `join()` without `select()` in a relationship query will fill the fetched Eloquent model with attributes from secondary table(s)
2. `join()` may require `select()` disambiguating columns for `id`, etc. if the relationship query is reused.
    * e.g., `$user->postsOfManyLocations()->someScope()->get()`
3. `join()` can't be applied through multiple query scopes without secondary joins needing a named alias, limiting scope reuse.

Traditionally SQL joins were encouraged over subqueries but the latter really hasn't been a performance bottleneck in basic web apps since the early 2000s. In Laravel codebases, joins are really only something I've used in one-off reporting queries when I do want the results to reflect multiple tables or aggregates. When relationships & scopes became join-heavy, it mainly exposed Eloquent pain points. To fill in the gaps, devs needed to do some extra work (including custom query macros) not covered by official Laravel documentation.

**Comment by taylorotwell** [comment]
Thanks for looking into this @derekmd

**Comment by theriddleofenigma** [comment]
@derekmd Using eloquent relation like has() won't solve the need here. Since @troy-whitespark tries to load the count of locations mapped against a post. And so he used joins in the relationship.

@troy-whitespark I have added a PR for solving your need, #41359. Kindly check and share your thoughts on this.

**Comment by derekmd** [comment]
```php
User::with([
    'posts' => function ($query) {
        $query->withCount('locations');
    },
])->get();
```

```php
$user->posts->loadCount('locations');
```

**Comment by theriddleofenigma** [comment]
@derekmd Yeah, we can achieve it using this. But it will result in 1 additional query. If we achieve it using the `withCount()` itself, then we can save a query. Also, that will reduce the iterations for hydrating the post relation. Since we don't need the **posts** details here.

**Comment by taylorotwell** [comment]
Since there is a work around available for solving this issue and it isn't a fundamental flaw in the framework that is super critical at the moment, I'm going to close this for now since we don't have any plans to look into it internally (we use open issues to see what we actually plan to tackle ourselves). Of course, anyone else is welcome to take a stab at it if they want.
